File: src/http.m
// HTTPS via NSURLSession.
//
// Not std.http.Client: the standard library's HTTP and TLS surface has changed
// shape across recent Zig releases, and this project already carries the ObjC
// bridge pattern (local_auth.m, lockstate.m) with dispatch_semaphore to make an
// async API synchronous. It also keeps request bodies out of argv, which
// shelling out to curl would not — the bodies here are signed approval
// requests, and `ps` is readable by any process on the machine.
//
// Build flags in build.zig: -fobjc-arc + -framework Foundation.

#import <Foundation/Foundation.h>
#include <string.h>

// Return codes. Negative means the request never completed, so the caller can
// tell "the network failed" from "the service said no", which are different
// things when the answer gates a secret.
#define SC_HTTP_OK             0
#define SC_HTTP_BAD_ARGS      -1
#define SC_HTTP_TRANSPORT     -2
#define SC_HTTP_TIMEOUT       -3
#define SC_HTTP_BODY_TOO_BIG  -4

/// Perform one request and block until it finishes.
///
/// `headers` is a NUL-terminated blob of "Key: Value\n" lines — one string
/// rather than a marshalled array, because the Zig side builds it once and this
/// avoids two allocations and a length-array per call.
///
/// The response body is written into a caller-provided buffer. Nothing is
/// allocated across the boundary, so there is no free() to forget.
int secretctl_http(const char *method,
                   const char *url,
                   const char *headers,
                   const unsigned char *body,
                   size_t body_len,
                   int timeout_ms,
                   int *out_status,
                   unsigned char *out_buf,
                   size_t out_cap,
                   size_t *out_len) {
    if (!method || !url || !out_status || !out_len) return SC_HTTP_BAD_ARGS;
    *out_status = 0;
    *out_len = 0;

    @autoreleasepool {
        NSURL *nsurl = [NSURL URLWithString:[NSString stringWithUTF8String:url]];
        if (!nsurl) return SC_HTTP_BAD_ARGS;
        // Refuse plaintext outright. An http:// endpoint would put signed
        // requests and sealed bodies on the wire in clear, and the only way
        // that URL gets here is a mistake or an attacker-supplied config.
        if (![[nsurl.scheme lowercaseString] isEqualToString:@"https"]) {
            return SC_HTTP_BAD_ARGS;
        }

        NSMutableURLRequest *req = [NSMutableURLRequest requestWithURL:nsurl];
        req.HTTPMethod = [NSString stringWithUTF8String:method];
        req.timeoutInterval = (double)timeout_ms / 1000.0;
        // Approval requests must never be answered from a cache: a cached
        // verdict is a replayed verdict.
        req.cachePolicy = NSURLRequestReloadIgnoringLocalAndRemoteCacheData;

        if (headers && headers[0]) {
            NSString *blob = [NSString stringWithUTF8String:headers];
            for (NSString *line in [blob componentsSeparatedByString:@"\n"]) {
                if (line.length == 0) continue;
                NSRange colon = [line rangeOfString:@": "];
                if (colon.location == NSNotFound) continue;
                NSString *k = [line substringToIndex:colon.location];
                NSString *v = [line substringFromIndex:colon.location + colon.length];
                [req setValue:v forHTTPHeaderField:k];
            }
        }

        if (body && body_len > 0) {
            req.HTTPBody = [NSData dataWithBytes:body length:body_len];
        }

        NSURLSessionConfiguration *cfg =
            [NSURLSessionConfiguration ephemeralSessionConfiguration];
        cfg.timeoutIntervalForRequest = (double)timeout_ms / 1000.0;
        cfg.timeoutIntervalForResource = (double)timeout_ms / 1000.0;
        // No cookies, no credential store, no shared cache: this client has one
        // identity and it is a signature, not ambient session state.
        cfg.HTTPCookieStorage = nil;
        cfg.URLCache = nil;
        NSURLSession *session = [NSURLSession sessionWithConfiguration:cfg];

        // Everything the caller needs is copied out inside the completion
        // handler, before the semaphore is signalled. Nothing Objective-C
        // outlives the handler.
        //
        // The earlier shape kept `__block NSData *` alive across the wait and
        // read it afterwards, which is correct only because ARC retains on
        // assignment — so its correctness depended on -fobjc-arc being passed.
        // A probe build that omitted the flag segfaulted on the second request.
        // Copying here removes the dependency instead of documenting it.
        __block int status = 0;
        __block size_t copied = 0;
        __block int result = SC_HTTP_TRANSPORT;
        dispatch_semaphore_t sem = dispatch_semaphore_create(0);

        NSURLSessionDataTask *task =
            [session dataTaskWithRequest:req
                      completionHandler:^(NSData *d, NSURLResponse *r, NSError *e) {
                          if (e) {
                              result = (e.code == NSURLErrorTimedOut)
                                  ? SC_HTTP_TIMEOUT : SC_HTTP_TRANSPORT;
                          } else if (![r isKindOfClass:[NSHTTPURLResponse class]]) {
                              result = SC_HTTP_TRANSPORT;
                          } else {
                              status = (int)((NSHTTPURLResponse *)r).statusCode;
                              size_t n = d ? (size_t)d.length : 0;
                              if (n > out_cap) {
                                  result = SC_HTTP_BODY_TOO_BIG;
                              } else {
                                  if (n > 0 && out_buf) memcpy(out_buf, d.bytes, n);
                                  copied = n;
                                  result = SC_HTTP_OK;
                              }
                          }
                          dispatch_semaphore_signal(sem);
                      }];
        [task resume];

        // Bounded wait, with headroom over the request timeout so a hung
        // request surfaces as a timeout rather than blocking forever. The
        // unbounded DISPATCH_TIME_FOREVER in local_auth.m is exactly the bug
        // this avoids repeating.
        long waited = dispatch_semaphore_wait(
            sem, dispatch_time(DISPATCH_TIME_NOW,
                               (int64_t)(timeout_ms + 2000) * NSEC_PER_MSEC));
        if (waited != 0) {
            [task cancel];
            [session invalidateAndCancel];
            return SC_HTTP_TIMEOUT;
        }

        *out_status = status;
        *out_len = copied;
        [session finishTasksAndInvalidate];
        return result;
    }
}

File: src/http.zig
//! Thin Zig wrapper over the NSURLSession bridge in `http.m`.
//!
//! The response buffer is caller-owned and nothing crosses the boundary
//! allocated, so there is no free() to forget on either side.

const std = @import("std");

extern "c" fn secretctl_http(
    method: [*:0]const u8,
    url: [*:0]const u8,
    headers: [*:0]const u8,
    body: ?[*]const u8,
    body_len: usize,
    timeout_ms: c_int,
    out_status: *c_int,
    out_buf: ?[*]u8,
    out_cap: usize,
    out_len: *usize,
) c_int;

pub const Error = error{
    BadArgs,
    Transport,
    Timeout,
    ResponseTooLarge,
    OutOfMemory,
};

pub const Response = struct {
    status: u16,
    /// Owned by the caller-supplied allocator.
    body: []u8,

    pub fn deinit(self: Response, allocator: std.mem.Allocator) void {
        allocator.free(self.body);
    }

    pub fn ok(self: Response) bool {
        return self.status >= 200 and self.status < 300;
    }
};

pub const Request = struct {
    method: []const u8 = "GET",
    url: []const u8,
    /// "Key: Value\n" lines. Built by the caller; see headerList.
    headers: []const u8 = "",
    body: []const u8 = "",
    /// Must exceed the service's hold (25 s) for the waiting call, or every
    /// wait would look like a timeout.
    timeout_ms: c_int = 15_000,
    /// Approval payloads are small envelopes; a cap keeps a hostile or broken
    /// service from being able to make this allocate without bound.
    max_response: usize = 256 * 1024,
};

pub fn send(allocator: std.mem.Allocator, req: Request) Error!Response {
    const method_z = allocator.dupeZ(u8, req.method) catch return Error.OutOfMemory;
    defer allocator.free(method_z);
    const url_z = allocator.dupeZ(u8, req.url) catch return Error.OutOfMemory;
    defer allocator.free(url_z);
    const headers_z = allocator.dupeZ(u8, req.headers) catch return Error.OutOfMemory;
    defer allocator.free(headers_z);

    const buf = allocator.alloc(u8, req.max_response) catch return Error.OutOfMemory;
    errdefer allocator.free(buf);

    var status: c_int = 0;
    var out_len: usize = 0;
    const rc = secretctl_http(
        method_z,
        url_z,
        headers_z,
        if (req.body.len > 0) req.body.ptr else null,
        req.body.len,
        req.timeout_ms,
        &status,
        buf.ptr,
        buf.len,
        &out_len,
    );

    switch (rc) {
        0 => {},
        -1 => return Error.BadArgs,
        -3 => return Error.Timeout,
        -4 => return Error.ResponseTooLarge,
        else => return Error.Transport,
    }

    // Shrink to what arrived. The caller frees whatever comes back.
    const body = allocator.realloc(buf, out_len) catch buf[0..out_len];
    return .{ .status = @intCast(status), .body = body };
}

/// Build a "Key: Value\n" blob. Caller owns the result.
pub fn headerList(
    allocator: std.mem.Allocator,
    pairs: []const [2][]const u8,
) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(allocator);
    for (pairs) |kv| {
        // A newline or colon smuggled into a header value would let a caller
        // inject additional headers. Values here are ids and base64url, so
        // rejecting is right and silently stripping would hide a bug.
        if (std.mem.indexOfAny(u8, kv[1], "\r\n") != null) return error.BadHeaderValue;
        try out.appendSlice(allocator, kv[0]);
        try out.appendSlice(allocator, ": ");
        try out.appendSlice(allocator, kv[1]);
        try out.append(allocator, '\n');
    }
    return out.toOwnedSlice(allocator);
}

test "headerList formats and rejects injected newlines" {
    const a = std.testing.allocator;

    const h = try headerList(a, &.{
        .{ "X-Nudge-Client", "abc" },
        .{ "content-type", "application/json" },
    });
    defer a.free(h);
    try std.testing.expectEqualStrings(
        "X-Nudge-Client: abc\ncontent-type: application/json\n", h);

    try std.testing.expectError(error.BadHeaderValue, headerList(a, &.{
        .{ "X-Nudge-Sig", "abc\r\nX-Nudge-Client: someone-else" },
    }));
}

test "plaintext http is refused before any connection is attempted" {
    const a = std.testing.allocator;
    // Signed requests and sealed bodies must never go out in clear, and the
    // only way such a URL reaches here is a mistake or a tampered config.
    try std.testing.expectError(Error.BadArgs, send(a, .{
        .url = "http://example.invalid/v1/health",
        .timeout_ms = 1000,
    }));
}
